lib/quicly.c

```diff
@@ -3378,6 +3378,27 @@ static int do_send(quicly_conn_t *conn, quicly_send_context_t *s)
             goto Exit;
     }
 
+    /* handling PTO: send an ACK-eliciting packet if something is inflight */
+    if (restrict_sending) {
+        if (conn->egress.last_retransmittable_sent_at == now) {
+            /* we have committed at least one ACK-eliciting frame */
+        } else if (s->target.packet != NULL) {
+            /* we have a packet being built, which is not ACK-eliciting. Convert it to ACK-eliciting. */
+            assert(!s->target.ack_eliciting);
+            assert(s->dst < s->dst_end);
+            *s->dst++ = QUICLY_FRAME_TYPE_PING;
+            s->target.ack_eliciting = 1;
+            conn->egress.last_retransmittable_sent_at = now;
+        } else if (conn->egress.sentmap.bytes_in_flight != 0 || conn->super.peer.address_validation.send_probe) {
```

Comment:
This "if" can be changed to an assert, as it is the prerequisite of the PTO timer firing (though we need to think about the case we are sending address validation probes.